pkg/failure/event_forwarding.go

```diff
@@ -0,0 +1,157 @@
+	processorArtifact string
+	processorVersion  string
+	target            targetiface.Target
+	log               *log.Entry
+}
+
+// NewEventForwardingFailure will create a new client for handling failed events
+// by converting them into EventForwarding compatible bad events and pushing them to
+// a stream
+func NewEventForwardingFailure(target targetiface.Target, processorArtifact string, processorVersion string) (*EventForwardingFailure, error) {
+	return &EventForwardingFailure{
+		processorArtifact: processorArtifact,
+		processorVersion:  processorVersion,
+		target:            target,
+		log:               log.WithFields(log.Fields{"failed": EventForwardingFailureTarget}),
+	}, nil
+}
+
+// WriteInvalid will handle the conversion of invalid messages into failure
+// messages that will then pushed to the specified target
+func (ef *EventForwardingFailure) WriteInvalid(invalid []*models.Message) (*models.TargetWriteResult, error) {
+	var transformed []*models.Message
+
+	for _, msg := range invalid {
+		var failureErrors []string
+
+		err := msg.GetError()
+		if err != nil {
+			failureErrors = append(failureErrors, err.Error())
+		}
+
+		var sv *badrows.BadRow
+		reportableError, ok := err.(models.ErrorMetadata)
+		if ok {
+			sv, err = badrows.NewEventForwardingError(
+				&badrows.EventForwardingErrorInput{
+					ProcessorArtifact: ef.processorArtifact,
+					ProcessorVersion:  ef.processorVersion,
+					OriginalTSV:       msg.OriginalData,
+					ErrorType:         reportableError.ReportableType(),
+					LatestState:       msg.Data,
+					ErrorMessage:      reportableError.ReportableDescription(),
+					ErrorCode:         reportableError.ReportableCode(),
+					FailureTimestamp:  msg.TimePulled,
+				},
+				ef.target.MaximumAllowedMessageSizeBytes(),
+			)
+		} else {
+			sv, err = badrows.NewEventForwardingError(
+				&badrows.EventForwardingErrorInput{
+					ProcessorArtifact: ef.processorArtifact,
+					ProcessorVersion:  ef.processorVersion,
+					OriginalTSV:       msg.OriginalData,
+					LatestState:       msg.Data,
+					ErrorMessage:      strings.Join(failureErrors, ": "),
+					FailureTimestamp:  msg.TimePulled,
+				},
+				ef.target.MaximumAllowedMessageSizeBytes(),
+			)
+		}
+
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to transform invalid message to event_forwarding.event_forwarding_error bad-row JSON")
+		}
+
+		svCompact, err := sv.Compact()
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to get compacted event_forwarding.event_forwarding_error bad-row JSON")
+		}
+
+		tMsg := msg
+		tMsg.Data = []byte(svCompact)
+
+		transformed = append(transformed, tMsg)
+	}
+
+	return ef.target.Write(transformed)
+}
+
+// WriteOversized will handle the conversion of oversized messages into failure
+// messages that will then pushed to the specified target
+func (ef *EventForwardingFailure) WriteOversized(maximumAllowedSizeBytes int, oversized []*models.Message) (*models.TargetWriteResult, error) {
```

Comment:
nit - bit of repeated code here but I understand that it might be more hassle than it's worth to refactor things to not repeat it.

If you see a simple way to address it give it a crack but I'm fine with it the way it is